### ISSUE TITLE
Cannot to attach the remote dlv process

### ISSUE THREAD

**Issue by Karmenzind** [leoluz/nvim-dap-go#35]
Environment:

```
dlv version:
Delve Debugger
Version: 1.20.1
Build: $Id: 96e65b6c615845d42e0e31d903f6475b0e4ece6e $

nvim-dap-go is the latest version
commit b4ded7de579b4e2a85c203388233b54bf1028816 (HEAD -> main, origin/main, origin/HEAD)

/proc/version:
Linux version 6.1.6-arch1-3 (linux@archlinux) (gcc (GCC) 12.2.1 20230111, GNU ld (GNU Binutils) 2.40) #1 SMP PREEMPT_DYNAMIC Mon, 16 Jan 2023 12:51:23 +0000

Lua config:
require("dap-go").setup({
    dap_configurations = {
        { type = "go", name = "Attach remote", mode = "remote", request = "attach" },
    },
})
```

![image](https://user-images.githubusercontent.com/17923071/217781272-80fd1f38-b403-4421-978d-8ebd8bddecd3.png)

Steps to reproduce:
1. Run my gin project with `dlv debug -l 127.0.0.1:38697 --headless xxx/xxx.go`
2. `7. Attach Remote` raised `Error on attach: Failed to attach: no debugger found`
3. Run `4. Attach` and input 111 (`111: id=1332739 name=dlv debug -l ...` in the process list) and it raised `Error on attach: Failed to attach: could not attach to pid 1332739: could not open debug info - debuggee must not be built with 'go run' or -ldflags='-s -
w', which strip debug info`

BTW `1. Debug` worked fine but it surpressed the output.

I don't know which part went wrong. Several days ago (before I update my system) I just need to run `dlv debug -l 127.0.0.1:38697 --headless xxx/xxx.go`and select `4. Attach` and it would automatically find the remote process and attach to it.

**Comment by Kasama** [comment]
I'm having the same issue here. Did you get it to work? Some months ago I'm sure it worked, but now I'm getting this `Error on attach: Failed to attach: no debugger found`.

`:lua = require('dap').configurations.go` shows
```lua
{
  {
    name = "Debug",
    program = "${file}",
    request = "launch",
    type = "go"
  }, {
    args = <function 1>,
    name = "Debug (Arguments)",
    program = "${file}",
    request = "launch",
    type = "go"
  }, {
    name = "Debug Package",
    program = "${fileDirname}",
    request = "launch",
    type = "go"
  }, {
    mode = "local",
    name = "Attach",
    processId = <function 2>,
    request = "attach",
    type = "go"
  }, {
    mode = "test",
    name = "Debug test",
    program = "${file}",
    request = "launch",
    type = "go"
  }, {
    mode = "test",
    name = "Debug test (go.mod)",
    program = "./${relativeFileDirname}",
    request = "launch",
    type = "go"
  }, {
    mode = "remote",
    name = "Attach to remote Delve",
    request = "attach",
    type = "go"
  }
}
```

and `:lua = require('dap').adapters.go`
```lua
{
  executable = {
    args = { "dap", "-l", "127.0.0.1:${port}" },
    command = "dlv"
  },
  options = {
    initialize_timeout_sec = 20
  },
  port = "${port}",
  type = "server"
}
```

**Comment by Kasama** [comment]
Nevermind, i think this is a dup of #27.

Adding the following made it work for me
```lua
require('dap-go').setup {
  delve = {
    port = "38697",
  },
}
```

**Comment by marcomayer** [comment]
This also stopped working for me (debugger not found) with the setup from the README:
 ```
delve = {
    -- time to wait for delve to initialize the debug session.
    -- default to 20 seconds
    initialize_timeout_sec = 20,
    -- a string that defines the port to start delve debugger.
    -- default to string "${port}" which instructs nvim-dap
    -- to start the process in a random available port
    port = "${port}"
  },
```

For some reason dropping `initialize_timeout_sec` fixes it:
 ```
delve = {
    port = "${port}"
  },
```

**Comment by marcomayer** [comment]
> I had this issue to: I added this to my config to resolve my issue

This is really helpful. I have the problem too with the launching dlv instances and solved it differently by stopping them afterwards but this is much better!

**Comment by manofthelionarmy** [comment]
> Only, issue now is that if you want to debug a test for example it doesn't work anymore.

Yeah I figured that'll break that😵😞. Hence, a separate adapter for headless debug servers. Apart from that, I had issues with the attach. Though the go adapter is configured to start the server, the process doesn't get kicked off and I don't see it's pid. And, if I do start a dlv process in headless mode, and attach to its pid via nvim dap, I get an error saying the debug session is already in progress at port 8181 and instead use remote mode to connect to the server. I wonder if the go adapter for attach should be setup as an executable, which can start the server, and we'll see it's pid from nvim dap.

**Comment by marcomayer** [comment]
What works for me right now is setting up nvim-dap like this:

```
	require("dap-go").setup({
		delve = {
			port = "38697",
		},
		dap_configurations = {
			{
				type = "go",
				name = "Attach remote",
				mode = "remote",
				request = "attach",
					},
		},
	})
```

And starting the debug server with:
`while pkill dlv; dlv debug -l 127.0.0.1:38697 --headless ./cmd/main.go; do echo "restarting delve..."; done`

So it kills off the dlv started by nvim locally first. Of course this is suboptimal but I didn't have the time to really look into this. But that way both work, attaching to remote or starting one up for test debugging etc.

**Comment by leoluz** [comment]
> There haven't been any commits and issues/PRs look stale though not sure if this plugin is still maintained @leoluz ?

@marcomayer Yes it is and I use it often :) I work on it when I have time.
I am also open for PRs and help from the community to do issue triage.

**Comment by psethwick** [comment]
FWIW, this config works totally fine for me without the pkill hack:
```
 require('dap-go').setup {
  dap_configurations = {
    {
      type = "go",
      name = "Attach remote",
      mode = "remote",
      request = "attach",
      connect = {
        host = "127.0.0.1",
        port = "43000"
      }
    },
  },
  delve = {
    port = "43000"
  },
}
```
Debug test works fine, attach to remote works fine.

**Comment by leoluz** [comment]
I don't use the remote debug feature and I am not sure if there is still an issue with clashing dlv instances running in the same port. 

Can someone please confirm so I can close this issue?

**Comment by psethwick** [comment]
I haven't seen it, as one data point.

It wasn't a perfect test, but I just now started a remote and connected to it with htop open (filtered to 'dlv') and didn't see any extras pop up.

**Comment by paulbrittain** [comment]
@psethwick 's config worked for me in-so-far as the connection was made to the remote Delve. However, the listener couldn't be started because the port is in use. 

```
require('dap-go').setup {
  -- Additional dap configurations can be added.
  -- dap_configurations accepts a list of tables where each entry
  -- represents a dap configuration. For more details do:
  -- :help dap-configuration
  dap_configurations = {
    {
      -- Must be "go" or it will be ignored by the plugin
      type = "go",
      name = "Attach remote",
      mode = "remote",
      request = "attach",
    },
    {
      type = "go",
      name= "Service-1",
      request= "attach",
      mode= "remote",
    }
  },
  delve = {
      port = 45000,
  }
}
```

`couldn't start listener: listen tcp 127.0.0.1:45000: bind: address already in use`

I have a number of containerized Go services, and each can run delve on a diferent port. IDEs such as Vscode (config below) and Goland can attach to these processes

VSCode debug config:
```
"configurations": [
        {
            "name": "Service-1",
            "port": 45000,
            "type": "go",
            "request": "attach",
            "mode": "remote",
            "showLog": true,
            "trace": "log",
            "logOutput": "rpc"
        },
        {
            "name": "Service-2",
            "port": 35000,
            "type": "go",
            "request": "attach",
            "mode": "remote",
            "showLog": true,
            "trace": "log",
            "logOutput": "rpc"
        },
    ]
   ```
   
   Ideally, the dap_configuration would handle different ports for different configuration entries too.
   
   Any help welcome! If I get this working I'll be sure to report back here.
 
   
   ```
  dap_configurations = {
    {
      -- Must be "go" or it will be ignored by the plugin
      type = "go",
      name = "Attach remote",
      mode = "remote",
      request = "attach",
    },
    {
      type = "go",
      name= "Service-1",
      port= 45000,
      request= "attach",
      mode= "remote",
    },
    {
      type = "go",
      name= "Service-2",
      --port= 35000,
      request= "attach",
      mode= "remote",
      port = "35000"
    }
  },
}
   
   ```

**Comment by leoluz** [comment]
@paulbrittain Thank you for your investigation. I am closing this for now. Please feel free to reopen if the issue persists.

**Comment by stevenctl** [comment]
fwiw I'm getting the exact same issue. seems like somehow the plugin is both trying to attach _and_ start it's own headless dlv process? 

When I attach I get a warning that `dlv exited with code 1` but the UI and logs from my debug process show that I'm attached. Breakpoints do not pause, the repl doesn't give any accurate info.

**Comment by jsternberg** [comment]
@stevenctl I'm running into the same issue. I was able to track it down to likely this commit: https://github.com/leoluz/nvim-dap-go/commit/bd9823da22165ea4aa7ec232e8c3a9b3158e017a

There was a custom delve adapter with a callback and it started using the builtin `nvim-dap` adapter, but the `nvim-dap` adapter doesn't seem to look at the `request` type before it determines what it does. I tracked it to this call [here](https://github.com/mfussenegger/nvim-dap/blob/6f79b822997f2e8a789c6034e147d42bc6706770/lua/dap.lua#L439). It calls `attach` which I thought was a strange name, but this then calls `connect` [here](https://github.com/mfussenegger/nvim-dap/blob/6f79b822997f2e8a789c6034e147d42bc6706770/lua/dap/session.lua#L1300) and the connect method seems to always launch the executable if the executable is set.

The only way I can see to fix it is to use the `enrich_config` to exclude the executable when the request is `attach` in this plugin or potentially modify this in `nvim-dap`.

I just started using nvim maybe last week though so I'm not sure where to make the change upstream or how to modify this locally. @leoluz I'd like to make a PR for this if I can get some guidance on where the change should be made. I use remote debugging pretty frequently by attaching to a delve server inside of a docker container.
